### PR TITLE
[libomptarget][OpenMP] Initial implementation of omp_target_memset() and omp_target_memset_async()

### PR DESCRIPTION
**Pull request by mjklemm**
This PR adds the newly landed OpenMP API routines omp_target_memset() and omp_target_memset_sync() to fill memory in the device memory. This first implementation is based on a slow path that initializes memory on the host and then issues a H2D transfer to fill the memory on the target device. A future PR will introduce an offloaded kernel (or driver-side API to do fill the memory on the target device)

### CODE REVIEW

File: openmp/libomptarget/include/omptarget.h [hunk 1]
```diff
@@ -312,6 +312,7 @@ int omp_target_memcpy_rect(void *Dst, const void *Src, size_t ElementSize,
                            const size_t *DstDimensions,
                            const size_t *SrcDimensions, int DstDevice,
                            int SrcDevice);
+void *omp_target_memset(void *Ptr, int C, size_t N, int DeviceNum);
```
Comment (shiltian) [inline openmp/libomptarget/include/omptarget.h:315]: More descriptive argument name for `C` and `N`.

File: openmp/libomptarget/src/api.cpp [hunk 2]
```diff
@@ -241,10 +241,125 @@ static int libomp_target_memcpy_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
   return Rc;
 }
 
+static int libomp_target_memset_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
+  if (!Task) {
+    return OFFLOAD_FAIL;
+  }
```
Comment (shiltian) [inline openmp/libomptarget/src/api.cpp:247]: LLVM style please (no `{}` for single statement.

File: openmp/libomptarget/src/private.h [hunk 3]
```diff
@@ -253,6 +253,19 @@ struct TargetMemcpyArgsTy {
         DstOffsets(DstOffsets), SrcOffsets(SrcOffsets),
         DstDimensions(DstDimensions), SrcDimensions(SrcDimensions){};
 };
+
+struct TargetMemsetArgsTy {
+  /**
```
Comment (shiltian) [inline openmp/libomptarget/src/private.h:258]: style

File: openmp/libomptarget/src/api.cpp [hunk 4]
```diff
@@ -241,10 +241,125 @@ static int libomp_target_memcpy_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
...
+  (void)libomp_helper_memset_task_creation(Args, DepObjCount, DepObjList);
+
+  return Ptr;
+}
+
 // Allocate and launch helper task
-static int libomp_helper_task_creation(TargetMemcpyArgsTy *Args,
-                                       int DepObjCount,
-                                       omp_depend_t *DepObjList) {
+static int libomp_helper_memcpy_task_creation(TargetMemcpyArgsTy *Args,
```
Comment (shiltian) [inline openmp/libomptarget/src/api.cpp:360]: Is it possoble to make this interface generic to make the implementation of future async APIs easier? For example, we might be able to pass a callback for the parts that are different for different APIs.

File: openmp/libomptarget/include/omptarget.h (hunk 1 above)
Comment (mjklemm) [reply openmp/libomptarget/include/omptarget.h:315]: I can do that.  I have "borrowed" the argument names from the `memset` interface: `void *memset(void *s, int c, size_t n);`, but I'm OK with being a bit more verbose and use more elaborate names.

File: openmp/libomptarget/src/api.cpp (hunk 2 above)
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp:247]: Ack!

File: openmp/libomptarget/src/api.cpp (hunk 4 above)
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp:360]: I have thought about this and I will see if I can refactor the code a bit to improve this and get rid of the two implementations.

File: openmp/libomptarget/src/api.cpp (hunk 2 above)
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp]: Resolved via commit 7933bfa0.

File: openmp/libomptarget/include/omptarget.h (hunk 1 above)
Comment (mjklemm) [reply openmp/libomptarget/include/omptarget.h:315]: Resolved via commit 7933bfa0.

File: openmp/libomptarget/src/api.cpp (hunk 4 above)
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp]: How about the last patch set?  The duplication of the helper function, but I have retained the actual task function.  There's still a tiny bit of code duplication, but I felt like that the reduction using a callback is not saving much at the expense of making the actual control flow even harder to understand.  Let me know if you'd agree to that,

Comment (shiltian) [reply openmp/libomptarget/src/api.cpp]: Thanks for the effort. I'm in LLVM Dev Meeting. I'll take a look afterwards.

File: openmp/libomptarget/src/api.cpp [hunk 5]
```diff
@@ -241,47 +241,130 @@ static int libomp_target_memcpy_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
...
-// Allocate and launch helper task
-static int libomp_helper_task_creation(TargetMemcpyArgsTy *Args,
-                                       int DepObjCount,
-                                       omp_depend_t *DepObjList) {
+static int libomp_target_memset_async_task(kmp_int32 Gtid, kmp_task_t *Task) {
+  if (!Task)
+    return OFFLOAD_FAIL;
+
+  auto *Args = reinterpret_cast<TargetMemsetArgsTy *>(Task->shareds);
+  if (!Args) {
```
[the remark this answers is not in the document]
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp:249]: Curly braces are gone (in the next round of pushing my patch set).

File: openmp/libomptarget/src/private.h (hunk 3 above)
Comment (shiltian) [reply openmp/libomptarget/src/private.h:258]: Usually the comment for a class is outside of the class

File: openmp/libomptarget/src/api.cpp [hunk 6]
```diff
@@ -241,47 +241,130 @@ static int libomp_target_memcpy_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
...
+                    omp_depend_t *DepObjList) {
+  for (int i = 0; i < DepObjCount; ++i) {
+    omp_depend_t DepObj = DepObjList[i];
+    Vec.push_back(*((kmp_depend_info_t *)DepObj));
+  }
+}
+
+template <class T>
+static inline int
+libomp_helper_task_creation(T *Args, int (*Fn)(kmp_int32, kmp_task_t *),
```
Comment (shiltian) [inline openmp/libomptarget/src/api.cpp:272]: I think it would work if `void *Args` since `Args` is assigned to `shareds`.

**Review by shiltian — COMMENTED** [review]
The rest of the patch looks good to me, though I'm not sure why I still see some old version of the code.

**Comment by mjklemm** [comment]
I have pushed a the latest patch set now.  Does it look better now?

File: openmp/libomptarget/src/api.cpp (hunk 6 above)
Comment (mjklemm) [reply openmp/libomptarget/src/api.cpp:271]: The exact type is needed for the `delete` in line 285.  So, I'd prefer to leave the templated code.

File: openmp/libomptarget/src/private.h (hunk 3 above)
Comment (mjklemm) [reply openmp/libomptarget/src/private.h]: Resolved this one via a regular comment.

**Comment by mjklemm** [comment]
> The rest of the patch looks good to me, though I'm not sure why I still see some old version of the code.

If you feel that this is good to, please merge (I cannot, as I have not yet earned the power to do so)

File: openmp/libomptarget/include/omptarget.h (hunk 1 above)
Comment (shiltian) [reply openmp/libomptarget/include/omptarget.h:315]: I'm kinda lost actually. On my side (the "Files changed" tab), I still see `C` and `N`, but you have apparently updated it. Other stuff is similar. I'm not sure what caused this.

Comment (mjklemm) [reply openmp/libomptarget/include/omptarget.h:315]: I guess it's a GitHub PR thing.  The review comments are still attached to the old code that the comment was targeting.  I guess you'd have to go to the "Files Changed" tab of the PR to see the latest changes and get a chance to create a new comment thread.  I'll mark this one here as "resolved" for now.

File: openmp/libomptarget/src/api.cpp [hunk 7]
```diff
@@ -241,47 +241,129 @@ static int libomp_target_memcpy_async_helper(kmp_int32 Gtid, kmp_task_t *Task) {
...
+  // call omp_target_memset()
+  omp_target_memset(Args->Ptr, Args->C, Args->N, Args->DeviceNum);
+
+  delete Args;
+
+  return OFFLOAD_SUCCESS;
+}
+
+static inline void
+ConvertDepObjVector(llvm::SmallVector<kmp_depend_info_t> &Vec, int DepObjCount,
```
Comment (jplehr) [inline openmp/libomptarget/src/api.cpp:261]: Shouldn't this function name be lower case?

**Review by jplehr — APPROVED** [review]
LGTM with one nit

**Pull request merged**